### PR TITLE
[WIP] Information filter and Lyapunov filter

### PR DESCRIPTION
**Pull request by frankschae**
This adds the information filter of Theorem 3.3.  Per default, we stilll use the covariance filter of Theorem 3.4. By using the `filter=` kwarg, we may however now switch between the two.  We may next easily add also the LyapunovFilter of Proposition 3.5 -- that probably needs a better name though (?) 

I'll still have to add more tests and also the `convert_message` function.  

@fmeulen

### CODE REVIEW

**Comment by mschauer** [comment]
Perhaps it comes in handy that you can just write
```
convert(WGaussian{(:F,:Γ,:c)}, ...)
```
to convert between different parametrisations.

File: test/mitosis_test.jl [hunk 1]
```diff
@@ -75,11 +75,10 @@ m, p = Mitosis.backwardfilter(gkernel, y)
 WG = WGaussian{(:μ,:Σ,:c)}(y_, 0.0, 0.0) #
 
 message, solend = MitosisStochasticDiffEq.backwardfilter(sdekernel, WG)
-
 @testset "Mitosis backward" begin
-    @test (p.c)[] ≈ solend.c
-    @test (p.Γ\p.F)[] ≈ solend.μ atol=atol
-    @test inv(p.Γ)[] ≈ solend.Σ atol=atol
+    @test (p.y.c)[] ≈ solend.c
```
[the remark this answers is not in the document]
Comment (mschauer) [reply test/mitosis_test.jl:79]: Perhaps https://github.com/mschauer/Mitosis.jl/pull/10

[the remark this answers is not in the document]
Comment (frankschae) [reply test/mitosis_test.jl]: Possibly, it's worth to include a downstream test in Mitosis

[the remark this answers is not in the document]
Comment (mschauer) [reply test/mitosis_test.jl]: I have to figure that out... I have fixed things, everything works and this PR does not conflict with the changes.

**Comment by frankschae** [comment]
Added the analytical filter based on the Lyapunov equation. Potentially there is a small error, in the solution for `\nu` -- That one doesn't look good, see the Figure below, where the difference with respect to the ODE solution is plotted. 

difference

File: src/filter.jl [hunk 2]
```diff
@@ -89,3 +172,54 @@ function backwardfilter(k::SDEKernel, (c, ν, P)::NamedTuple{(:logscale, :μ, :
...
+Φfunc(t,T,B) = exp(-(T-t)*B)
+function solP(t,T,B,PT,Σ)
+  Φ = Φfunc(t,T,B)
+  Φ*(Σ+PT)*Φ'-Σ
+end
+
+function solν(t,T,B,β,νT)
+  #TODO
+  # β time-dependent
+  νT*Φfunc(t,T,-B) + Φfunc(t,T,-B)*β*inv(B)*(Φfunc(t,T,B)-1)
```
Comment (frankschae) [inline src/filter.jl:186]: It would be good to double check this line.

File: src/guiding.jl [hunk 3]
```diff
@@ -66,19 +70,24 @@ function (G::GuidingDriftCache)(u,p,t)
...
   end
 
-  r = P\(ν .- x)
-
-  dl = dot(f(x,p,t) -  ktilde.f(x,ktilde.p,t), r) - 0.5*tr((outer_(g(x,p,t)) - outer_(ktilde.g(x,ktilde.p,t)))*(inv(P) - outer_(r)))
-  dx = vec(f(x, p, t) + outer_(g(x, p, t))*r) # evolution guided by observations
+  if !(filter isa InformationFilter)
+    r = Σ\(μ .- x)
 
+    dl = dot(f(x,p,t) -  ktilde.f(x,ktilde.p,t), r) - 0.5*tr((outer_(g(x,p,t)) - outer_(ktilde.g(x,ktilde.p,t)))*(inv(Σ) - outer_(r)))
```
Comment (frankschae) [inline src/guiding.jl:84]: Is `dl` written down in https://arxiv.org/abs/1712.03807? Couldn't immediately find it.

Comment (mschauer) [reply src/guiding.jl:84]: You mean equation (2.5)?

File: src/filter.jl (hunk 2 above)
Comment (mschauer) [reply src/filter.jl:186]: I have a time-inhomogeneous example in https://github.com/mschauer/Bridge.jl/blob/430e9d8a0a39a2cd00f13cd10bfd1f3df5f3d2c4/src/linpro.jl#L13 Essentially we could try to reproduce tests which are using this https://github.com/mschauer/Bridge.jl/search?q=Ptilde

**Pull request merged**